client/network/src/service.rs

```diff
@@ -1281,637 +1280,633 @@ where
-				Poll::Ready(SwarmEvent::Behaviour(BehaviourOut::InboundRequest {
-					protocol,
-					result,
-					..
-				})) => {
-					if let Some(metrics) = this.metrics.as_ref() {
-						match result {
-							Ok(serve_time) => {
-								metrics
-									.requests_in_success_total
-									.with_label_values(&[&protocol])
-									.observe(serve_time.as_secs_f64());
-							},
-							Err(err) => {
-								let reason = match err {
-									ResponseFailure::Network(InboundFailure::Timeout) => "timeout",
-									ResponseFailure::Network(
-										InboundFailure::UnsupportedProtocols,
-									) =>
-									// `UnsupportedProtocols` is reported for every single
-									// inbound request whenever a request with an unsupported
-									// protocol is received. This is not reported in order to
-									// avoid confusions.
-										continue,
-									ResponseFailure::Network(InboundFailure::ResponseOmission) =>
-										"busy-omitted",
-									ResponseFailure::Network(InboundFailure::ConnectionClosed) =>
-										"connection-closed",
-								};
+		// Next swarm event, or `future::pending()` if the stream has terminated
+		// (guaranteed to never happen with `Swarm`).
+		let next_swarm_event = {
+			let swarm = &mut self.network_service;
 
-								metrics
-									.requests_in_failure_total
-									.with_label_values(&[&protocol, reason])
-									.inc();
-							},
-						}
+			async move {
+				if let Some(event) = swarm.next().await {
+					event
+				} else {
+					loop {
+						futures::pending!();
 					}
-				},
-				Poll::Ready(SwarmEvent::Behaviour(BehaviourOut::RequestFinished {
-					protocol,
-					duration,
-					result,
-					..
-				})) =>
-					if let Some(metrics) = this.metrics.as_ref() {
-						match result {
-							Ok(_) => {
-								metrics
-									.requests_out_success_total
-									.with_label_values(&[&protocol])
-									.observe(duration.as_secs_f64());
-							},
-							Err(err) => {
-								let reason = match err {
-									RequestFailure::NotConnected => "not-connected",
-									RequestFailure::UnknownProtocol => "unknown-protocol",
-									RequestFailure::Refused => "refused",
-									RequestFailure::Obsolete => "obsolete",
-									RequestFailure::Network(OutboundFailure::DialFailure) =>
-										"dial-failure",
-									RequestFailure::Network(OutboundFailure::Timeout) => "timeout",
-									RequestFailure::Network(OutboundFailure::ConnectionClosed) =>
-										"connection-closed",
-									RequestFailure::Network(
-										OutboundFailure::UnsupportedProtocols,
-									) => "unsupported",
-								};
+				}
+			}
+		};
 
-								metrics
-									.requests_out_failure_total
-									.with_label_values(&[&protocol, reason])
-									.inc();
-							},
-						}
+		futures::select! {
+			msg = next_worker_msg.fuse() => {
+				// Process the next message coming from the `NetworkService`.
+				match msg {
+					ServiceToWorkerMsg::AnnounceBlock(hash, data) => self
```

Comment:
This whole method's really long and the code's nested very deeply; could we perhaps refactor it a little bit while we're at it? e.g. this part (and the next one for the swarm events) which matches the message and handles it, can we move it into a separate method and then just simply call `self.handle_worker_message(msg)` here?

Reply:
Done :+1: